cmd/flat-network-controller/app/controller.go

```diff
@@ -0,0 +1,247 @@
+
+	if err := manageFlatNetwork(net, c.netClientset, c.svcClientset); err != nil {
+		c.queue.AddRateLimited(key)
+		c.recorder.Eventf(net, corev1.EventTypeWarning, "FailedProvision", "failed to provision network %s/%s/%s: %v", net.Tenant, net.Tenant, net.Name, err)
+		return
+	}
+
+	c.recorder.Eventf(net, corev1.EventTypeNormal, "SuccessfulProvision", "successfully provision network %s/%s/%s", net.Tenant, net.Tenant, net.Name)
+	c.queue.Forget(item)
+}
+
+// manageFlatNetwork is the core logic to manage a flat typed network object
+func manageFlatNetwork(net *v1.Network, netClient arktos.Interface, svcClient kubernetes.Interface) error {
+	if len(net.Status.DNSServiceIP) != 0 {
+		klog.V(5).Infof("network %s/%s/%s already have DNS service IP %s; skipped", net.Tenant, net.Namespace, net.Name, net.Status.DNSServiceIP)
+		return nil
+	}
+
+	svc, err := createOrGetDNSService(net, svcClient)
+	if err != nil {
+		return err
+	}
+
+	// since dns service is in place, flat type network is Ready now
+	netReady := net.DeepCopy()
+	netReady.Status.Phase = v1.NetworkReady
+	netReady.Status.DNSServiceIP = svc.Spec.ClusterIP
+	netReady.Status.Message = "DNS service ready; network ready"
+	_, err = netClient.ArktosV1().NetworksWithMultiTenancy(netReady.Tenant).UpdateStatus(netReady)
+	return err
+}
+
+func createOrGetDNSService(net *v1.Network, svcClient kubernetes.Interface) (*corev1.Service, error) {
+	const nsDNS = "kube-system"
+	nameDNS := "kube-dns-" + net.Name
+	dns := corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      nameDNS,
+			Tenant:    net.Tenant,
+			Namespace: nsDNS,
+		},
+		Spec: corev1.ServiceSpec{
+			Ports: []corev1.ServicePort{
+				{
+					Name:       "dns",
+					Protocol:   "UDP",
+					Port:       53,
+					TargetPort: intstr.FromInt(53),
+				},
+				{
+					Name:       "dns-tcp",
+					Protocol:   "TCP",
+					Port:       53,
+					TargetPort: intstr.FromInt(53),
+				},
+				{
+					Name:       "metrics",
+					Protocol:   "TCP",
+					Port:       9153,
+					TargetPort: intstr.FromInt(9153),
+				},
+			},
+			Selector: map[string]string{
+				"k8s-app": "kube-dns",
+			},
+			Type: "ClusterIP",
+		},
+	}
+	svc, err := svcClient.CoreV1().ServicesWithMultiTenancy(nsDNS, net.Tenant).Create(&dns)
+	if err != nil {
+		if !errors.IsAlreadyExists(err) {
+			// todo: consider differing temporary errors from permanent errors
+			// todo: to fail the network provision in case of permanent errors
+			return nil, err
+		} else {
+			svc, err = svcClient.CoreV1().ServicesWithMultiTenancy(nsDNS, net.Tenant).Get(nameDNS, metav1.GetOptions{})
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	return svc, nil
+}
+
+// Enqueue puts key of the network object in the work queue
+func (c *Controller) Enqueue(obj interface{}) {
+	net, ok := obj.(*v1.Network)
+	if !ok {
+		klog.Fatalf("got non-network object %v", obj)
+	}
+
+	c.queue.Add(genKey(net))
+}
+
+func genKey(net *v1.Network) string {
```

Comment:
let's use cache.MetaNamespaceKeyFunc(), the systematic to generate cache key.

Reply:
better not to depend on internal mechanism. It is purposed as controller/resource specific way for custom keys, though coincidently very similar to that used in apiserver in this case..